pandas/core/groupby/groupby.py

```diff
@@ -691,7 +691,14 @@ def _set_result_index_ordered(self, result):
             result.set_axis(index, axis=self.axis, inplace=True)
             result = result.sort_index(axis=self.axis)
 
-        result.set_axis(self.obj._get_axis(self.axis), axis=self.axis, inplace=True)
+        if (
+            self.dropna
+            and self.axis == 0
+            and len(result.index) < len(self._selected_obj.index)
+        ):
+            result.index = self._selected_obj.index[result.index.asi8]
```

Comment:
`index.asi8` is often None, e.g. for a MultiIndex. Removing this line (but not the if-else blocks!) also passes all groupby tests for me. Recommend finding what tests fail without the call to set_axis below, and seeing if you an replicate the failure without this line in a dropna example.

Reply:
Did as suggested. Groupby tests still pass, but now so do the master versions of the transform tests. Thanks!!!

(That was on my local toolchain. Will see if CI turns something up)

Reply:
In the else block here, we are using self._selected_obj's index rather than the result's index (the latter I think is always a range index). It seems to me we need to do that same operation in the case some rows are dropped because of nans. I think it's likely that there are tests for (a) non-standard indices or (b) dropna, but not both. Can you check to see if dropna behaves well when the index is not a rangeindex?

Reply:
Yes the problem is there like you say:
```
In [7]: df = pd.DataFrame({"A": [0, 0, 1, None], "B": [1, 2, 3, None]}, index=list('abcd')) 
   ...: gb = df.groupby('A', dropna=False) 
   ...: result = gb["B"].transform(len)                                                                                                                    

In [8]: result                                                                                                                                             
Out[8]: 
a    2
b    2
c    1
d    1
Name: B, dtype: int64

In [9]: df = pd.DataFrame({"A": [0, 0, 1, None], "B": [1, 2, 3, None]}, index=list('abcd')) 
   ...: gb = df.groupby('A', dropna=True) 
   ...: result = gb["B"].transform(len)                                                                                                                    

In [10]: result                                                                                                                                            
Out[10]: 
0    2
1    2
2    1
Name: B, dtype: int64
```
Fixed this is in the new commit & added more testcases (a non-standard simple index like above and a MultiIndex) which will now work